common_util/ast_util.py

```diff
@@ -0,0 +1,121 @@
+import ast
+from dataclasses import dataclass
+from typing import List
+
+from common_util.const import NOT_A_MODEL_NAME, ML_MODELS_LIST
+
+
+def get_func_name(node: ast.Call) -> str:
+    # Assume that 'foo' can't be model name in case of such Call: foo(...)(...)
+    # Only in case of such Call: foo(...)
+    if not isinstance(node, ast.Call):
+        raise TypeError(f'Expected ast.Call node, but {type(node).__name__} received')
+
+    try:
+        return node.func.id
+    except AttributeError:
+        try:
+            return node.func.attr
+        except AttributeError:
+            return NOT_A_MODEL_NAME
+
+
+def continue_traverse(traverse, node, cell_no):
+    for field_name in node._fields:
+        field = getattr(node, field_name)
+        if isinstance(field, list):
+            for elem in field:
+                traverse(elem, cell_no)
+        traverse(field, cell_no)
+
+
+@dataclass
+class NtbTokenLocationInfo:
+    cell_no: int = None
+    line_no: int = None
+
+
+@dataclass
+class NtbModelInfo:
+    model_name: str = None
+    variable_name: str = None
+    constructor_location: NtbTokenLocationInfo = None
+    fit_location: NtbTokenLocationInfo = None
+    X_train_var_name: str = None
+    y_train_var_name: str = None
+
+
+class NtbModelInfoAstTraverser:
+    model_info: List[NtbModelInfo]
+
+    def __init__(self):
+        self.model_info = []
+
+    def traverse_collect_model_info(self, node: ast.AST, cell_no: int):
+        if not isinstance(node, ast.AST):
+            return
+
+        candidate_node_predicate = isinstance(node, ast.Assign) and \
+                                   isinstance(node.value, ast.Call) and \
+                                   get_func_name(node.value) in ML_MODELS_LIST + ['fit']
+
+        if candidate_node_predicate:
+
+            call_node = node.value
+            func_name = get_func_name(call_node)
+
+            if func_name in ML_MODELS_LIST:
+                ntb_model_info = self.__get_model_info_for_model_node(node, func_name, cell_no)
+
+            else:  # func_name == 'fit'
+                ntb_model_info = self.__get_model_info_for_fit_node(call_node, node, cell_no)
+
+            if ntb_model_info is not None:
+                self.model_info.append(ntb_model_info)
+        else:
+            continue_traverse(self.traverse_collect_model_info, node, cell_no)
+
+    def traverse_fill_missing_fit_loc(self, node: ast.AST, cell_no: int):
+        if not isinstance(node, ast.AST):
+            return
+
+        missing_fit_loc_models = [mi for mi in self.model_info if mi.fit_location is None]
+
+        if len(missing_fit_loc_models) == 0:
+            return
+
+        if isinstance(node, ast.Call) and get_func_name(node) == 'fit':
+            try:
```

Comment:
Most simple example is when we have function call like `a = fit(...)`. Here fit is not a method of some ML model object, but separate function which maybe defined by user of something). In that case everything inside try would fail, obviously.